app/scripts/controller-init/types.ts

```diff
@@ -0,0 +1,168 @@
+import { Provider } from '@metamask/network-controller';
+import {
+  ActionConstraint,
+  ControllerMessenger,
+  EventConstraint,
+  RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import { TransactionMetricsRequest } from '../lib/transaction/metrics';
+import { Controller, ControllerFlatState } from './controller-list';
+
+/** The supported controller names. */
+export type ControllerName = Controller['name'];
+
+/** All controller types by name. */
+export type ControllerByName = {
+  [name in ControllerName]: Controller & { name: name };
+};
+
+/**
+ * Persisted state for all controllers.
+ * e.g. `{ TransactionController: { transactions: [] } }`.
+ */
+export type ControllerPersistedState = {
+  [name in ControllerName]: ControllerByName[name]['state'];
```

Comment:
Not all state is persisted, so `Partial` seems appropriate here

```suggestion
  [name in ControllerName]: Partial<ControllerByName[name]['state']>;
```

In theory we could derive a more correct representation of persisted state using the controller metadata, but that's just set at runtime at the moment. Not statically in the type. Plus any old installation might have a different set of state persisted, so `Partial` is maybe a safer bet anyway.

Reply:
Done.